[ADD] stock_location_orderpoint_average_daily_sale

rousseldenis opened the pull request:
Depends on:

- #3 
- https://github.com/OCA/stock-logistics-reporting/pull/257

rousseldenis committed:
[ADD] stock_location_orderpoint_average_daily_sale

rousseldenis committed:
[ADD] stock_location_orderpoint_average_daily_sale

rousseldenis committed:
[DONT MERGE] test-requirements.txt

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Perf

Compute recommended qty for all products at once by warehouse

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Perf

Compute recommended qty for all products at once by warehouse

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Perf tests

Initialize common test data into setupClass and add dedicated tests for cron

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Perf tests

Initialize common test data into setupClass and add dedicated tests for cron

jbaudoux commented on stock_location_orderpoint_average_daily_sale/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,101 @@
+        # need value, the existing quantity on destination and the already replenished
+        # quantity
+        needed_quantity = (
+            recommended_qty - virtual_available_on_dest - qty_already_replenished
```
```suggestion
            recommended_qty - max(virtual_available_on_dest, 0) - qty_already_replenished
```

jbaudoux reviewed (approved):
LG, tested

rousseldenis committed:
[IMP] stock_location_orderpoint_average_daily_sale: Improved _compute_quantities_dict in base module

rousseldenis committed:
Revert "[IMP] stock_location_orderpoint_average_daily_sale: Improved _compute_quantities_dict in base module"

This reverts commit 86eb3d9378f72fef8aa5e7105d6e7fb71a5ad41f.

rousseldenis committed:
[FIX] stock_location_orderpoint_average_daily_sale: Make the quantities computation more robust

rousseldenis committed:
[FIX] stock_location_orderpoint_average_daily_sale: Make the quantities computation more robust

jbaudoux commented on stock_location_orderpoint_average_daily_sale/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,103 @@
+    _inherit = "stock.location.orderpoint"
+
+    replenish_method = fields.Selection(
+        selection_add=[("average_daily_sale", "Average Daily Sale")],
+        ondelete={"average_daily_sale": "cascade"},
```
I know it relies on `average_daily_sale` module but it's about past-looking average daily usage.
Can we at least drop the "sale" term that is misleading?
The module will be renamed in v18 to remove the confusion
```suggestion
        selection_add=[("average_daily_sale", "Past Average Daily Usage")],
        ondelete={"average_daily_sale": "cascade"},
```

jbaudoux commented on stock_location_orderpoint_average_daily_sale/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,103 @@
+        return res
+
+    def _add_recommended_qty_to_qties_dict(self, qties_on_locations, products):
+        # we add by warehouse the recommended_qty for each product
+        wh_ids = {location.warehouse_id.id for location in qties_on_locations}
+        recommended_qty = {wh_id: {} for wh_id in wh_ids}
+        qties_on_locations["recommended_qty"] = recommended_qty
+        product_ids = {p.id for p in products}
+        product_set = set(product_ids)
+        for wh_id in wh_ids:
+            not_processed_product_ids = product_set
+            reports = self.env["stock.average.daily.sale"].search(
+                [
+                    ("product_id", "in", list(product_set)),
+                    ("warehouse_id", "=", wh_id),
+                ]
+            )
```
The avg config is per location, not per warehouse. You can compute avg on different locations inside a warehouse.
This needs to be changed and rely on the location. Maybe you want to configure on the location orderpoint the avg config location to use to retrieve metrics.

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Implements strategy

Refafactor the code to implement the strategy of average daily sale in
order to compute the minimum quantity of the orderpoint. This change
follows the refactoring introduced in https://github.com/OCA/stock-logistics-orderpoint/pull/85

lmignon commented:
ping @jbaudoux @rousseldenis

lmignon committed:
[FIX] stock_location_orderpoint_average_daily_sale: Demand must only contains product in daily report table

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Add replenish_horizon

 When computing a replenishment, it is important to exclude from the outgoing quantities the quantities that will themselves be replenished by a future replenishment.

    When replenishment is performed on a regular basis, only outgoing moves occurring before the next planned replenishment should be taken into account.

lmignon committed:
[FIX] stock_location_orderpoint_average_daily_sale: Excludes incoming moves from the demand calculation

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Add product domain to average daily usage strategy

lmignon commented:
rebased to include missing unmerged dependencies and make tests :green_circle:

mt-software-de commented on stock_location_orderpoint_average_daily_sale/models/stock_location_orderpoint.py:
```diff
@@ -0,0 +1,57 @@
+            return self.env["stock.location.orderpoint.strategy.average_daily_usage"]
+        return super()._strategy_model
+
+    # -------------------------------------------------------------------------
+    # Orderpoint selection and domains
+    # -------------------------------------------------------------------------
+
+    @api.model
+    @tools.ormcache("orderpoint_id")
+    def _get_daily_usage_products_location_domain(self, orderpoint_id):
+        """
+        In the average daily sale strategy, we want to consider all the products that
+        have an average daily usage at the destination location of the orderpoint as
+        candidates to compute the demand for.
+
+        :param orderpoint_id: int id of the orderpoint
+
+        :return: list of tuples to use as domain to filter products
+        """
+        orderpoint = self.browse(orderpoint_id)
+        (quant_domain, _il, _ol,) = orderpoint._product_model._get_domain_locations_new(
+            [orderpoint.location_id.id]
+        )
+        return extract_subdomains(quant_domain, "location_id")
+
+    # -------------------------------------------------------------------------
+    # Cache lifecycle
+    # -------------------------------------------------------------------------
+
+    def _clear_caches(self):  # pylint: disable=missing-return
+        super()._clear_caches()
+        self._get_daily_usage_products_location_domain.clear_cache(self)
```
`_get_daily_usage_products_location_domain` of stock.location.orderpoint isn't used.

mt-software-de reviewed (changes requested):
THX. Looks good. Just a minor request.

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Refactor orderpoint to be location specific

divad1196 commented:
@rousseldenis @lmignon  dependencies are merged,

We just need to remove commit [2286f30850fbb1a9d746144b96bacf22e7e96e35: [DONT MERGE] test-requirements.txt
](https://github.com/OCA/stock-logistics-orderpoint/pull/4/changes/2286f30850fbb1a9d746144b96bacf22e7e96e35) and it can be merged

lmignon commented:
@divad1196 The initial dependencies are merged but not the new one on https://github.com/OCA/stock-logistics-orderpoint/pull/85

divad1196 commented:
@lmignon I didn't notice it was a different one. I thought it was the 2 from the description
sorry :pray:

lmignon commented:
> @lmignon I didn't notice it was a different one. I thought it was the 2 from the description sorry 🙏

no stress :smile:

lmignon committed:
[IMP] stock_location_orderpoint_average_daily_sale: Rename cache clearing method for clarity and consistency

lmignon commented:
@jbaudoux This one is ready to be merged. :smirk:

jbaudoux commented:
/ocabot merge nobump
